File: perl/csv2md.pl
#!/usr/bin/perl

# Converts a CSV (comma-separated values) formatted table into a
# MD (markdown) formatted equivalent.
#
# The CSV-formatted lines is expected to come from STDIN. The MD-formatted
# output is sent to STDOUT.
#
# If comma is used as a delimiter, both command lines are valid:
#     cat input.csv | ./csv2md.pl > output.md
#     cat input.csv | ./csv2md.pl ',' > output.md
#
# If semi-colon is used, then this should be specified instead:
#     cat input.csv | ./csv2md.pl ';' > output.md

use strict;
use warnings;
use FileHandle;

STDOUT->autoflush;

sub report_error
{
    my $error_message = shift;
    print STDERR $error_message . "\r\n";
    exit 1;
}

sub program_info
{
    my $info = "Converts a CSV (comma-separated values) formatted table into a " . 
               "MD (markdown) formatted equivalent.\r\n\r\n" .
               "The CSV-formatted lines is expected to come from STDIN. The " .
               "MD-formatted output is sent to STDOUT.\r\n\r\n" .
               "If comma is used as a delimiter, both command lines are valid:\r\n" .
               "\tcat input.csv | ./csv2md.pl > output.md\r\n" .
               "\tcat input.csv | ./csv2md.pl ',' > output.md\r\n\r\n" .
               "If semi-colon is used, then this should be specified instead:\r\n" .
               "\tcat input.csv | ./csv2md.pl ';' > output.md";
    return $info;
}

# Standard delimiter
my $delimiter = ',';

my $nr_of_arguments = scalar(@ARGV);
# Single argument?
if ($nr_of_arguments == 1)
{
    my $argument = lc $ARGV[0];
    if ($argument eq '--help')
    {
        print program_info() . "\r\n";
        exit 0;
    }
    else
    {
        # Assume it is delimiter
        $delimiter = $argument;
        if ($delimiter ne ";" && $delimiter ne ",")
        {
            report_error("Invalid delimiter! You must use comma or semicolon!");
        }
    }
}
# More than one argument...
elsif ($nr_of_arguments > 1)
{
    report_error("Invalid arguments list!\r\n\r\n" . program_info());
}

my $is_header_row = 1;
my $line;
foreach $line (<STDIN>)
{
    $line =~ tr/\r//d;
    $line =~ tr/\n//d;
    if ($line !~ /[[:print:]]/)
    {
        report_error("Non-printable characters!");
    }
    local $_ = $line;
    eval("s/$delimiter/|/g");
    print "|" . "$_" . "|\r\n";
    if ($is_header_row)
    {
        my $nr_of_delimiters = () = $line =~ /$delimiter/g;
        my $nr_of_columns = $nr_of_delimiters + 1;
        print "|", "---|" x $nr_of_columns, "\r\n";
    }
    $is_header_row = 0;
}

File: python/csv2md.py
#!/usr/bin/env python

"""Converts a CSV (comma-separated values) formatted table into a
MD (markdown) formatted equivalent.

The CSV-formatted lines is expected to come from STDIN. The MD-formatted output
is sent to STDOUT.

If comma is used as a delimiter, both command lines are valid:
    cat input.csv | ./csv2md.py > output.md
    cat input.csv | ./csv2md.py ',' > output.md

If semi-colon is used, then this should be specified instead:
    cat input.csv | ./csv2md.py ';' > output.md
"""

import sys

class ErrorHandler:
    def report(self, error_message):
        print(error_message, file=sys.stderr)
        sys.exit(1)

class ProcessArguments:
    def __init__(self, error_handler):
        self.error = error_handler
        self.help_text = None
        self.delimiter_char = None

    def _set_delimiter_char(self, delimiter_char):
        if delimiter_char in [';', ',']:
            self.delimiter_char = delimiter_char
        else:
            self.error.report('Invalid delimiter! You must use comma or semicolon!')

    def process(self, arguments):
        assert isinstance(arguments, list)
        del arguments[0]
        nr_of_arguments = len(sys.argv)
        if nr_of_arguments == 0:
            self._set_delimiter_char(',')
        elif nr_of_arguments == 1:
            argument = arguments[0].lower()
            if argument == '--help':
                self.help_text = program_info()
            else:
                self._set_delimiter_char(argument)
        else:
            self.error.report('Invalid arguments list!\n\n' + program_info())

    def help(self):
        return self.help_text

    def delimiter(self):
        return self.delimiter_char

class CSVToMarkdown:
    def __init__(self, delimiter_char, error_handler):
        self.error = error_handler
        assert delimiter_char in [';', ',']
        self.delimiter_char = delimiter_char
        # The first row is assumed to be the header row.
        self.is_header_row = True

    def convert(self, line):
        line = line.replace('\r', '').replace('\n', '')
        if not line.isprintable():
            self.error.report('Non-printable characters!')
            return None
        output = '|' + line.replace(self.delimiter_char, '|') + '|'
        if self.is_header_row:
            output += '\n|' + '---|' * (line.count(self.delimiter_char) + 1)
            self.is_header_row = False
        return output

def program_info():
    info =  'Converts a CSV (comma-separated values) formatted table into a ' +     \
            'MD (markdown) formatted equivalent.\n\n' +                             \
            'The CSV-formatted lines is expected to come from STDIN. The ' +        \
            'MD-formatted output is sent to STDOUT.\n\n' +                          \
            'If comma is used as a delimiter, both command lines are valid:\n' +    \
            '\tcat input.csv | ./csv2md.py > output.md\n' +                         \
            '\tcat input.csv | ./csv2md.py \',\' > output.md\n\n' +                 \
            'If semi-colon is used, then this should be specified instead:\n' +     \
            '\tcat input.csv | ./csv2md.py \';\' > output.md'
    return info

if __name__ == '__main__':
    error_handler = ErrorHandler()
    arguments = ProcessArguments(error_handler)
    arguments.process(sys.argv)
    if arguments.help() != None:
        print(arguments.help())
        sys.exit()
    converter = CSVToMarkdown(arguments.delimiter(), error_handler)
    for line in sys.stdin:
        converted_line = converter.convert(line)
        print(converted_line, flush=True)